Updated role to depend on drush deploy command.

markdorison opened the pull request:
## Description
Updated role to depend on drush deploy command.

Requires drush 10.3 or later and would likely be tagged as a 2.0 release.

## Motivation / Context
https://github.com/drush-ops/drush/pull/4359

## Testing Instructions / How This Has Been Tested
Can be tested on CHQ with an update to drush.

## Documentation
Updated defaults yaml.

markdorison committed:
Updated role to depend on drush deploy command.

Requires drush 10.3 or later.

https://github.com/drush-ops/drush/pull/4359

ctorgalson commented:
Looks good so far. 

If, we wanted to implement this _alongside_ the current logic and run them according to the environment (with each in an individual task or task file), we could.

Using `command` and Jinja2 filters, we'd [get the Drush version on the cli and compare it to a known value](https://github.com/ctorgalson/ansible-role-vim/blob/640a3d71116f37317d6320993515642146f59d12/tasks/setup.yml#L4-L15), and continue according to the result of the comparison.

(Incidentally, the `version` filter used to be called `version_compare` and seems to have vanished from the docs...but the tests pass on the tasks in the role I linked to, so :shrug:).

adamzimmermann commented:
This is the best overview I could find of the need to import config twice.

* https://www.drupal.org/project/config_split/issues/2962511

This issue also touches on the problem, but in a more round-about way.

* https://www.drupal.org/project/config_split/issues/2873521#comment-12342010

Just posting this here for now. We can discuss more and decide if it's worth bringing to the attention of the drush maintainers.

markdorison commented:
I created a CHQ PR to test this change: https://github.com/ChromaticHQ/chromatichq.com/pull/2430

It deployed on Tugboat as expected.

adamzimmermann commented on tasks/main.yml:
```diff
@@ -79,24 +79,6 @@
-# cache before the Drush cache to avoid fatal errors due to stale cache in the
-# service container.
-- name: clear drush cache.
-  shell: "{{ deploydrupal_drush_path }} cache-clear drush"
```
Do we no longer need this?

```
$ drush cr -h
Rebuild a Drupal 8 site.

This is a copy of core/rebuild.php. Additionally
 it also clears Drush cache and Drupal's render cache.
```

This seems to indicate that it still does something. I think?

Additionally, tutorials for creating Drush commands with drush:9^ indicate that a `drush cr drush` is needed still from what I'm finding online.

adamzimmermann replied:
Or perhaps we don't need this since we have a `drush cr` in the `drush deploy` command now?

markdorison replied:
[drush cr is not run by drush deploy](https://www.drush.org/deploycommand/). I will add this back in.

adamzimmermann commented on tasks/main.yml:
```diff
@@ -161,13 +121,6 @@
     - deploydrupal_npm_theme_build
     - deploydrupal_npm_theme_path is defined
 
-- name: clear caches.
```
Don't we still need this after theme files are rebuilt? If the cache was set between the commands above and before theme building finished, we could have the wrong things in the cache, right?

markdorison replied:
I suppose that is true. Another option: What if we rebuilt the theme first?

adamzimmermann replied:
I'm all for that. @agarzola any concerns?

agarzola replied:
I don’t see a problem with building the theme first. Go for it!

markdorison replied:
This seems like a big change for some reason (makes me nervous). Let's see how it goes!

adamzimmermann commented on tasks/main.yml:
```diff
@@ -117,40 +99,18 @@
-  when:
-    - deploydrupal_config_import
-
-- name: print drush output
```
Do we need this? It seems that this just duplicates output in many situations.

markdorison replied:
This is printing the output for the above drush command, which we removed.

markdorison committed:
Add back drush cc drush.

markdorison committed:
Moved theme build tasks earlier in the process.

Moved theme build tasks earlier so they would run before drush
cache-rebuild (prevents the need for a second cache rebuild).

markdorison commented on tasks/main.yml:
```diff
@@ -117,56 +118,18 @@
   when:
-    - not deploydrupal_site_install
-
-- name: run theme build tasks.
```
Moved these earlier in the process so that a second `drush cr` is not needed.

adamzimmermann commented:
Are going to try to merge this soon, or at least target this branch for some Tugboat builds to see what affect it has on speed? I assume that is was brought this back on our radar?

markdorison commented:
> Are going to try to merge this soon, or at least target this branch for some Tugboat builds to see what affect it has on speed? I assume that is was brought this back on our radar?

[CHQ has been using this branch](https://github.com/ChromaticHQ/chromatichq.com/blob/9c26e83b2ae1a5c20c1b3d1d60e4d3e9e85096f2/ansible/requirements.yml#L5) for the past couple of months to deploy on Tugboat. I'd like to monitor for at least a few more deployments since we have made additional changes.

agarzola reviewed (approved):
Looks good to me!
